riscv-sbi-prop.md

```diff
@@ -0,0 +1,290 @@
+Only individual hart can remove itself from supervisor mode. It can be
+moved to normal state only by sbi_hart_up function.
+
+Return the appropriate SBI error code.
+
+```
+u32 sbi_hart_state(unsigned long hartid)
+```
+Returns the RISCV_POWER_STATE for a specific hartid. This will help make
+kexec like functionality more robust.
+
+```
+void sbi_system_reset(u32 reset_type)
+```
+Reset the entire system.
+
+## Return error code Table:
+Here are the SBI return error codes defined.
+
+| Error Type               | Value  |
+| -------------------------|:------:|
+|  SBI_ERR_SUCCESS         |  0     |
+|  SBI_ERR_FAILURE         | -1     |
+|  SBI_ERR_NOT_SUPPORTED   | -2     |
+|  SBI_ERR_INVALID_PARAM   | -3     |
+|  SBI_ERR_DENIED          | -4     |
+|  SBI_ERR_INVALID_ADDRESS | -5     |
+
+
+## Power State
+A RISC-V core can exist in any of the following power states.
+
+| Power states         | Value  | Description                                  |
+| -------------------  |:------:|:---------------------------------------------|
+| RISCV_HART_ON        | 0      | Powered up & operational.                    |
+| RISCV_HART_STANDBY   | 1      | Powered up but at reduced energy <br> consumption WFI instruction can be used to achieve this state|
+| RISCV_HART_RETENTION | 2      | Deeper low power state. No reset <br> required but higher wakeup latency|
+| RISCV_HART_OFF       | 3      | Powered off. Reset of the core required after power restore|
+
+**_TODO_**:
+
+We probably also need deeper hart power states or system-level power state reporting
+at some point, and non-HART components (caches, memory, etc).  Likely aligned with
+platform specs, once those start to show up.
+
+## Implementation
+Currently, SBI is implemented as a part of BBL. There is a different SBI implementation
+available in coreboot as well.
+
+Alternatively, a separate open BSD/MIT licensed SBI project can be created that
+can be used by anybody to avoid this kind of SBI fragmentation. This project can
+generate both a firmware binary (to be executed directly in M mode) or a static library
+that can be used by different boot loaders. It will also help individual bootloaders
+to either work from M or S mode without a separate SBI implementation.
+
+## Conclusion
+This proposal is far from perfect and absolutely any suggestion is welcome.
+Obviously, there are many other functionalities that can be added to this proposal.
+However, I just wanted to start with something that is an incremental change at
+best to kick off the discussion. The aim here is to initiate a discussion that
+can lead to a robust SBI specification.
+
+## Reference:
+
+[1] http://infocenter.arm.com/help/topic/com.arm.doc.den0022d/Power_State_Coordination_Interface_PDD_v1_1_DEN0022D.pdf
+
+[2] https://github.com/riscv/riscv-sbi-doc/blob/master/riscv-sbi.md
+
+[3] https://github.com/riscv/riscv-sbi-doc/pull/9
+
+[4] https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md
+
+## Appendix
```

Comment:
The fact that you need to do this at all is a warning that you are walking into a trap.

Reply:
Agreed. Will remove this scheme.